clippy_lints/src/splitting_strings_at_newlines.rs

```diff
@@ -0,0 +1,89 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::source::snippet;
+use clippy_utils::ty::implements_trait;
+use rustc_ast::ast::LitKind;
+use rustc_errors::Applicability;
+use rustc_hir::{Expr, ExprKind};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::declare_lint_pass;
+
+use rustc_span::symbol::sym;
+
+declare_clippy_lint! {
+    /// ### What it does
+    ///
+    /// Checks for usages of `str.trim().split("\n")` and `str.trim().split("\r\n")`.
+    ///
+    /// ### Why is this bad?
+    ///
+    /// Hard-coding the line endings makes the code less portable. `str.lines` should be used instead.
+    ///
+    /// ### Example
+    /// ```no_run
+    /// "some\ntext\nwith\nnewlines\n".trim().split('\n');
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// "some\ntext\nwith\nnewlines\n".lines();
+    /// ```
+    #[clippy::version = "1.76.0"]
+    pub SPLITTING_STRINGS_AT_NEWLINES,
+    suspicious,
+    "splitting a trimmed string at hard-coded newlines"
+}
+
+declare_lint_pass!(SplittingStringsAtNewlines => [SPLITTING_STRINGS_AT_NEWLINES]);
+
+// Check if an expr is of a type that implements `Deref` with `Target = str`
+fn derefs_into_str(cx: &LateContext<'_>, expr: &'_ Expr<'_>) -> bool {
+    matches!(
+        cx.tcx.get_diagnostic_item(sym::Deref),
+        Some(def_id) if (
+            implements_trait(cx, cx.typeck_results().expr_ty(expr), def_id, &[])
+            && matches!(
+                cx.get_associated_type(cx.typeck_results().expr_ty_adjusted(expr), def_id, "Target"),
+                Some(target) if target.is_str()
+            )
+        )
+    )
+}
+
+impl LateLintPass<'_> for SplittingStringsAtNewlines {
+    fn check_expr(&mut self, cx: &LateContext<'_>, expr: &'_ Expr<'_>) {
+        // We're looking for `A.trim().split(B)`, where the type of `A` implements `Deref` with `Target =
+        // str` (e.g. a `str` literal or an expression returning `String`), and `B` is a `Pattern`
+        // that hard-codes a newline (either `"\n"` or `"\r\n"`). There are a lot of ways to specify a
+        // pattern, and this lint only checks the most basic ones: a `'\n'`, `"\n"`, and `"\r\n"`.
+
+        if expr.span.from_expansion() {
+            // Don't generate lints in code expanded from macros
+            return;
+        }
+
+        if let ExprKind::MethodCall(split_method_name, split_receiver, split_args, _) = expr.kind
+            && split_method_name.ident.as_str() == "split"
+            && split_args.len() == 1
```

Comment:
```suggestion
        if let ExprKind::MethodCall(split_method_name, split_receiver, [split_arg], _) = expr.kind
            && split_method_name.ident.as_str() == "split"
```
Nitpicky but you can avoid the length check + indexing by pattern matching on the slice directly

Reply:
That is exactly the nit-picking I came here for! Thanks, fixed.